File: solutions/python/algorithm/graph/Exercise2.py
from structure.dynamic.DynamicGraphADT import DynamicGraphADT
from structure.dynamic.DynamicSetADT import DynamicSetADT


def max_edge_cost_from(graph, v):
    if graph.is_empty():
        raise ValueError("El grafo está vacío.")

    vertices = graph.get_vertices()
    copia = copiar_conjunto(vertices)

    max_cost = None
    found = False

    while not copia.isEmpty():
        destino = copia.choose()
        copia.remove(destino)

        if graph.exists_edge(v, destino):
            cost = graph.edge_weight(v, destino)
            if (max_cost is None) or (cost > max_cost):
                max_cost = cost
                found = True

    if not found:
        raise ValueError(f"El vértice {v} no tiene aristas salientes.")

    return max_cost


def copiar_conjunto(original):
    copia = DynamicSetADT()
    temporal = DynamicSetADT()

    while not original.isEmpty():
        elem = original.choose()
        original.remove(elem)
        copia.add(elem)
        temporal.add(elem)

    while not temporal.isEmpty():
        elem = temporal.choose()
        temporal.remove(elem)
        original.add(elem)

    return copia


grafo = DynamicGraphADT()
grafo.add_vertex(1)
grafo.add_vertex(2)
grafo.add_vertex(3)
grafo.add_edge(1, 2, 10)
grafo.add_edge(1, 3, 25)

print(max_edge_cost_from(grafo, 1))  # → 25

File: solutions/javascript/src/algorithm/graph/Exercise2.js
import DynamicSetADT from "../../structure/dynamic/DynamicSetADT.js";
import DynamicGraphADT from "../../structure/dynamic/DynamicGraphADT.js";

function maxEdgeWeightFromVertex(graph, v) {
  if (graph.isEmpty()) {
    throw new Error("El grafo está vacío.");
  }

  const vertices = graph.getVertxs();
  const copia = copiarSet(vertices);

  let maxWeight = null;
  let found = false;

  while (!copia.isEmpty()) {
    const destino = copia.choose();
    copia.remove(destino);

    if (graph.existsEdge(v, destino)) {
      const peso = graph.edgeWeight(v, destino);
      if (maxWeight === null || peso > maxWeight) {
        maxWeight = peso;
        found = true;
      }
    }
  }

  if (!found) {
    throw new Error(`El vértice ${v} no tiene aristas salientes.`);
  }

  return maxWeight;
}

function copiarSet(original) {
  const copia = new DynamicSetADT();
  const aux = new DynamicSetADT();

  while (!original.isEmpty()) {
    const elem = original.choose();
    original.remove(elem);
    copia.add(elem);
    aux.add(elem);
  }

  while (!aux.isEmpty()) {
    const elem = aux.choose();
    aux.remove(elem);
    original.add(elem);
  }

  return copia;
}


const grafo = new DynamicGraphADT();
grafo.addVertx(1);
grafo.addVertx(2);
grafo.addVertx(3);
grafo.addEdge(1, 2, 15);
grafo.addEdge(1, 3, 30);

console.log(maxEdgeWeightFromVertex(grafo, 1)); // → 30
